datapackage_pipelines_measure/processors/add_twitter_resource.py

```diff
@@ -0,0 +1,285 @@
+
+
+def _get_mentions_for_entity_from_source(entity, start_date, end_date):
+    '''Get the mentions metric for entity type.'''
+    all_tweets_in_search = _get_twitter_search_results(entity,
+                                                       start_date,
+                                                       end_date)
+    return len(all_tweets_in_search)
+
+
+def _get_interactions_for_entity_from_source(entity, start_date, end_date):
+    def _count_indicators_of_interaction(list_of_tweets):
+        '''This methods iterates over a list of tweets, and returns a two-tuple
+        of total count of favorites, and total count of retweets. The method
+        makes sure it doesn't double-count stats, so it skips tweets that are
+        themselves retweets, and counts only original tweets'''
+        favorite_count, retweet_count = 0, 0
+        for tweet in list_of_tweets:
+            if getattr(tweet, 'retweeted_status', None):
+                continue
+            favorite_count += tweet.favorite_count
+            retweet_count += tweet.retweet_count
+        return favorite_count, retweet_count
+
+    entity_type = _get_entity_type(entity)
+    if entity_type == 'account':
+        '''
+        Get interaction for a twitter account entity. This is done by getting
+        all the account tweets during given date range, and aggregate the
+        measured 'interaction' metrics - favorites and retweets.
+        '''
+        tweets = _get_account_tweets(entity, start_date, end_date)
+    elif entity_type == 'hashtag':
+        '''
+        Get interactions for a hashtag entity. This is done by searching for
+        the hashtag between the given date range, and aggregate the measured
+        'interaction' metrics - favorites and retweets.
+        '''
+        tweets = _get_twitter_search_results(entity, start_date, end_date)
+    favorite_count, retweet_count = _count_indicators_of_interaction(tweets)
+    return sum([favorite_count, retweet_count])
+
+
+parameters, datapackage, res_iter = ingest()
+
+entity = parameters['entity']
+project_id = parameters['project_id']
+safe_entity = _get_safe_entity(entity)
+resource = {
+    'name': safe_entity,
+    'path': 'data/{}.json'.format(safe_entity)
+}
+
+entity_type = _get_entity_type(entity)
+
+# Followers
+# This is requested directly from the twitter api
+
+followers_count = None
+if entity_type == 'account':
+    user = _get_user_account_from_twitter_api(entity)
+    followers_count = user.followers_count
+
+
+# Mentions & Interactions
+# These are requested for specified (and limited) timeframes from the twitter
+# api, then added to the previous result to get a total since we started
+# collecting data for the entity.
+
+# Get last run entity values
+filter_object = {
+    'entity': entity,
+    'entity_type': entity_type,
+    'source': 'twitter',
+    'project_id': project_id
+}
+datastore = get_datastore()
+entity_last_run = datastore.get_latest_from_table(filter_object,
```

Comment:
Not sure where this comment belongs so I'm putting it here.

I think that a simpler solution would be to always query 3 times for the past three days (i.e. today, yesterday & the day before that).
Then you generate 3 separate records for each of these query results.

When writing to the DB, you use the timestamp (and probably entity, entity_type and source) as key for updating the DB.

This way you don't need to store any permanent state (which will be deleted once the container is re-deployed).
Also, I think that conceptually aggregation can (and should) be done separately, and not in the raw collector - either in a separate pipeline or in a simple SQL query.